### ISSUE TITLE
Add not working using withConverter 

### ISSUE THREAD

**Issue by garlicPasta** [googleapis/nodejs-firestore#924]
#### Environment details

  -  executed on firebase function 
  - `@google-cloud/firestore` version: 3.5 (as a depency of firebase-admin)

#### Steps to reproduce

```
  class Post {
    constructor(readonly title: string, readonly author: string) {}

    toString(): string {
      return this.title + ', by ' + this.author
    }
  }

  const postConverter = {
    toFirestore(p: Post){
      console.log('toFirestore got called')
      return {title: p.title, author: p.author}
    },
    fromFirestore(
      snapshot: firebase.firestore.QueryDocumentSnapshot,
    ): Post {
      const data = snapshot.data()
      return new Post(data.title, data.author)
    }
  }

  await firestore.firestore()
    .collection('posts')
    .withConverter(postConverter)
    .add(new Post("Foo", "123"))
}
```
 
Receiving the following error:

```
Error: Value for argument "data" is not a valid Firestore document. Couldn't serialize object of type "Post". Firestore doesn't support JavaScript objects with custom prototypes (i.e. objects that were created via the "new" operator).
    at Object.validateDocumentData (/srv/node_modules/@google-cloud/firestore/build/src/write-batch.js:617:15)
    at CollectionReference.add (/srv/node_modules/@google-cloud/firestore/build/src/reference.js:1881:23)
    at test (/srv/lib/index.js:54:10)
    at Layer.handle [as handle_request] (/srv/node_modules/express/lib/router/layer.js:95:5)
    at next (/srv/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/srv/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/srv/node_modules/express/lib/router/layer.js:95:5)
    at /srv/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/srv/node_modules/express/lib/router/index.js:335:12)
    at next (/srv/node_modules/express/lib/router/index.js:275:10)
```

So it seems like to me that the `toFirestore` doesnt get called. If i call `toFirebase` hardcoded it works.

Thanks!

**Comment by thebrianchen** [comment]
@garlicPasta  Thanks for reporting this bug! We should have this fixed by the next release.

**Comment by jessmorecroft** [comment]
Hi,

I've noticed that the converter is lost in a few places where new queries are created from old ones. This still retains the type but data returned is not converted so runtime errors ensue. Any use of where or limit for example, which, looking at the underlying code, calls "queryOptions.with" without propagating the converter, so the default is used per the with function's signature. If the converter is added at the end it works of course - perhaps this is the required pattern but still the types should reflect this if so - that is, new queries from where or limit should revert to the DocumentData type.

**Environment details**

* Executed locally against emulator. eg. `gcloud beta emulators firestore start --host-port localhost:8080`
* @google-cloud/firestore v3.5.0

**Steps to reproduce**

```
import {
  Firestore,
  FirestoreDataConverter,
  DocumentData,
} from '@google-cloud/firestore';
import { deepEqual } from 'assert';
process.env['FIRESTORE_EMULATOR_HOST'] = 'localhost:8080';

const firestore = new Firestore();

class Widget {
  constructor(private readonly obj: any) {}
  foo() {
    return this.obj;
  }
}

const widgetConverter: FirestoreDataConverter<Widget> = {
  toFirestore: (obj: Widget): DocumentData => {
    return obj.foo();
  },
  fromFirestore: (data: DocumentData): Widget => {
    return new Widget(data);
  },
};

const widgets = firestore.collection('widgets').withConverter(widgetConverter);

async function testFails1() {
  await widgets.doc('123').set(new Widget({ name: 'mywidget' }));
  const result = await widgets.limit(1).get();
  // blows up here with 'TypeError: doc.data(...).foo is not a function' since not converted to widget
  const objs = result.docs.map(doc => doc.data().foo());
  deepEqual(objs, [{ name: 'mywidget' }]);
}

async function testFails2() {
  await widgets.doc('123').set(new Widget({ name: 'mywidget' }));
  const result = await widgets.where('name', '==', 'mywidget').get();
  // blows up here with 'TypeError: doc.data(...).foo is not a function' since not converted to widget
  const objs = result.docs.map(doc => doc.data().foo());
  deepEqual(objs, [{ name: 'mywidget' }]);
}

async function testWorks1() {
  await widgets.doc('123').set(new Widget({ name: 'mywidget' }));
  const result = await widgets
    .limit(1)
    .withConverter(widgetConverter)
    .get();
  const objs = result.docs.map(doc => doc.data().foo());
  deepEqual(objs, [{ name: 'mywidget' }]);
}

async function testWorks2() {
  await widgets.doc('123').set(new Widget({ name: 'mywidget' }));
  const result = await widgets
    .where('name', '==', 'mywidget')
    .withConverter(widgetConverter)
    .get();
  const objs = result.docs.map(doc => doc.data().foo());
  deepEqual(objs, [{ name: 'mywidget' }]);
}

async function test() {
  await testWorks1();
  await testWorks2();
  await testFails1(); // fails here
  await testFails2();
}

test();
```

So basically I'd suggest any use of queryOptions.with to be reviewed to ensure that any converter is being correctly propagated.

Cheers

**Comment by thebrianchen** [comment]
@jessmorecroft  Thanks for reporting this issue with `queryOptions.with` along with the provided test cases! Sorry for the trouble, and I'll try to have a fix out shortly.

**Comment by jessmorecroft** [comment]
No worries at all. Great library by the way!

**Comment by thebrianchen** [comment]
@jessmorecroft This issue will be marked as closed once once the fix for `CollectionReference.add()` goes in, but the issue you reported has a corresponding fix at #931
